protocols/rendezvous/src/server.rs

```diff
@@ -109,192 +149,249 @@ pub enum Event {
+        &mut self,
+        peer_id: &PeerId,
+        connection_id: &ConnectionId,
+        _: &ConnectedPoint,
+        _: <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
+        _remaining_established: usize,
+    ) {
+        self.registrations
+            .unregister_connection(*peer_id, *connection_id)
     }
 
     fn inject_event(
         &mut self,
-        peer_id: PeerId,
-        connection: ConnectionId,
-        event: handler::InboundOutEvent,
+        peer: PeerId,
+        _: ConnectionId,
+        event: from_fn::OutEvent<Result<Option<InboundOutEvent>, Error>, Void, Void>,
     ) {
         let new_events = match event {
-            handler::InboundOutEvent::InboundEvent { id, message } => {
-                handle_inbound_event(message, peer_id, connection, id, &mut self.registrations)
+            from_fn::OutEvent::InboundEmitted(Ok(Some(InboundOutEvent::NewRegistration(
+                new_registration,
+            )))) => {
+                let (registration, expiry) = self.registrations.add(new_registration);
+                self.next_expiry.push(expiry);
+
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::PeerRegistered { peer, registration },
+                )]
+            }
+            from_fn::OutEvent::InboundEmitted(Ok(Some(InboundOutEvent::RegistrationFailed {
+                error,
+                namespace,
+            }))) => {
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::PeerNotRegistered {
+                        peer,
+                        error,
+                        namespace,
+                    },
+                )]
+            }
+            from_fn::OutEvent::InboundEmitted(Ok(Some(InboundOutEvent::Discovered {
+                cookie,
+                registrations,
+                previous_registrations,
+            }))) => {
+                self.registrations
+                    .cookies
+                    .insert(cookie, previous_registrations);
+
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::DiscoverServed {
+                        enquirer: peer,
+                        registrations,
+                    },
+                )]
+            }
+            from_fn::OutEvent::InboundEmitted(Ok(Some(InboundOutEvent::DiscoverFailed {
+                error,
+            }))) => {
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::DiscoverNotServed {
+                        enquirer: peer,
+                        error,
+                    },
+                )]
+            }
+            from_fn::OutEvent::InboundEmitted(Ok(Some(InboundOutEvent::Unregister(namespace)))) => {
+                self.registrations.remove(namespace.clone(), peer);
+
+                vec![NetworkBehaviourAction::GenerateEvent(
+                    Event::PeerUnregistered { peer, namespace },
+                )]
```

Comment:
A lot of code here is only present to make things observable from the outside. We don't really need to emit all these events in order for the protocol to be functional.